src/kokkos_abstraction.hpp

```diff
@@ -168,17 +172,79 @@ static struct LoopPatternTPTVR {
 // inner Kokkos::ThreadVectorRange
 static struct LoopPatternTPTTRTVR {
 } loop_pattern_tpttrtvr_tag;
+// Translates to an outer team policy
+static struct LoopPatternTeamOuter {
+} loop_pattern_team_outer_tag;
+// Translates to an outter team policy with an inner collapse over a combination of
+// thread/vecctor
+static struct LoopPatternTeamCollapse {
+} loop_pattern_team_collapse;
 // Used to catch undefined behavior as it results in throwing an error
 static struct LoopPatternUndefined {
 } loop_pattern_undefined_tag;
+// Translates to a Kokkos::TeamPolicy that collapse Nteams outer loops
+// with Nthread & Nvector inner loop collapses
+template <size_t Nteam, std::size_t Nthread, std::size_t Nvector>
+struct LoopPatternCollapse {};
+
+// trait to track if pattern requests any type of hierarchial parallelism
+template <typename Pattern, typename, typename T = void>
+struct LoopPatternTeam : std::false_type {
```

Comment:
True. Originally I was inspired by the openACC `#pragma acc loop gang/worker/vector collapse(n)` directives. I thought it might be useful to be able to specify exactly how to distribute the work for a specific loop, e.g.
```cpp
parthenon::par_for(LoopPatternCollapse<3,0,4>(),  "kernel", exec_space, ...)
```
and from that view all the `TP***` patterns are just special cases. But I also don't have use case that would justify supporting it.

Reply:
Wouldn't you get the same thing from
```c++
template <std::size_t NMiddle, std::size_t NInner> 
struct LoopPatternCollapse {
  static constexpr bool is_team_pattern = true;
  static constexpr std::size_t Nvector = NInner;
  static constexpr std::size_t Nthread = NMiddle;
};

using LoopPatternTPTVR = LoopPatternCollapse<0, 1>;
constexpr auto loop_pattern_tptvr_tag = LoopPatternTPTVR();

Reply:
I see what you mean now. Yes that is a lot neater